src/Processors/Sources/MongoDBSource.cpp

```diff
@@ -470,72 +494,50 @@ Chunk MongoDBSource::generate()
     if (all_read)
         return {};
 
-    MutableColumns columns(description.sample_block.columns());
-    const size_t size = columns.size();
-
-    for (const auto i : collections::range(0, size))
-        columns[i] = description.sample_block.getByPosition(i).column->cloneEmpty();
+    auto columns = sample_block.cloneEmptyColumns();
+    size_t size = columns.size();
 
     size_t num_rows = 0;
-    while (num_rows < max_block_size)
+    for (const auto & doc : cursor)
     {
-        auto documents = cursor.nextDocuments(*connection);
-
-        for (auto & document : documents)
+        for (auto idx : collections::range(0, size))
         {
-            if (document->exists("ok") && document->exists("$err")
-                && document->exists("code") && document->getInteger("ok") == 0)
-            {
-                auto code = document->getInteger("code");
-                const Poco::MongoDB::Element::Ptr value = document->get("$err");
-                auto message = static_cast<const Poco::MongoDB::ConcreteElement<String> &>(*value).value();
-                throw Exception(ErrorCodes::MONGODB_ERROR, "Got error from MongoDB: {}, code: {}", message, code);
-            }
-            ++num_rows;
+            auto & sample_column = sample_block.getByPosition(idx);
+            auto value = doc[sample_column.name];
 
-            for (const auto idx : collections::range(0, size))
+            if (!value || value.type() == bsoncxx::type::k_null)
             {
-                const auto & name = description.sample_block.getByPosition(idx).name;
+                insertDefaultValue(*columns[idx], *sample_column.column);
 
-                bool exists_in_current_document = document->exists(name);
-                if (!exists_in_current_document)
+                if (sample_column.type->isNullable())
```

Comment:
Won't we just insert `NULL` for Nullable column in `insertDefaultValue`? Why do we need explicitly set value in nullmap again?

Reply:
Yes, it will more shorter, thanks.